includes/misc-functions.php

```diff
@@ -795,3 +795,42 @@ function getallheaders() {
 	}
 
 endif;
+
+/**
+ * Determines the receipt visibility status
+ *
+ * @return bool Whether the receipt is visible or not.
+ */
+function edd_can_view_receipt( $payment_key = '' ) {
+
+	$return = false;
+
+	if ( empty( $payment_key ) ) {
+		return $return;
+	}
+
+	global $edd_receipt_args;
+
+	$edd_receipt_args['id'] = edd_get_purchase_id_by_key( $payment_key );
+
+	$user_id = edd_get_payment_user_id( $edd_receipt_args['id'] );
+
+	$payment_meta = edd_get_payment_meta( $edd_receipt_args['id'] );
+
+	if ( is_user_logged_in() && $user_id == get_current_user_id() ) {
+		$return = true;
+	}
+
+	if ( current_user_can( 'view_shop_sensitive_data' ) ) {
+		$return = true;
+	}
+
+	$session = edd_get_purchase_session();
+	if ( ! empty( $session ) && ! is_user_logged_in() ) {
```

Comment:
This should also have a condition for when the user ID on the payment doesn't match the the logged-in user's ID but the email does. This would happen where a payment was purchased while logged out (so no ID recorded) but with same email.